src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -169,6 +169,10 @@ public async Task SendRequestBodyAsync(CancellationToken cancellationToken)
                 {
                     if (NetEventSource.IsEnabled) Trace($"Failed to send request body: {e}");
 
+                    // Cancel the stream before we set _requestCompletionState below.
+                    // Otherwise, a response stream reader may race with the actual Cancel.
+                    Cancel();
+
                     lock (SyncObject)
                     {
                         Debug.Assert(_requestCompletionState == StreamCompletionState.InProgress, $"Request already completed with state={_requestCompletionState}");
```

Comment:
Is this assert still valid?  The Cancel couldn't cause the state to change?

Reply:
It's still valid. Cancel will not modify _requestCompletionState.